src/ORM/DataObject.php

```diff
@@ -2841,49 +2842,62 @@ public function setField($fieldName, $val)
             // later referenced to update the parent dataobject
             if ($val instanceof DBComposite) {
                 $val->bindTo($this);
-                $this->record[$fieldName] = $val;
+                $this->setFieldValue($fieldName, $val);
             }
         // Situation 2: Passing a literal or non-DBField object
         } else {
-            // If this is a proper database field, we shouldn't be getting non-DBField objects
-            if (is_object($val) && $schema->fieldSpec(static::class, $fieldName)) {
-                throw new InvalidArgumentException('DataObject::setField: passed an object that is not a DBField');
-            }
+            $this->setFieldValue($fieldName, $val);
+        }
+        return $this;
+    }
 
-            if (!empty($val) && !is_scalar($val)) {
-                $dbField = $this->dbObject($fieldName);
-                if ($dbField && $dbField->scalarValueOnly()) {
-                    throw new InvalidArgumentException(
-                        sprintf(
-                            'DataObject::setField: %s only accepts scalars',
-                            $fieldName
-                        )
-                    );
-                }
-            }
+    /**
+     * Usually you should call setField() instead
+     *
+     * Intended to called from setField(), or DBField::saveInto() which is called from setField()
+     * that used to use dynamic properties e.g. $dataObject->$fieldName = $value
+     * prior to dynamic properties being deprecationed in PHP 8.2
+     */
+    public function setFieldValue(string $fieldName, mixed $val): void
```

Comment:
Logic for this method was cut and pasted from setField()

Has the added benefit that when a DBField is passed as $val in setField(), $fieldName is now added to $this->changed whereas previously it was (incorrectly?) not